### PR TITLE
Document the kubernetes CoreDNS release procedure

### PR DESCRIPTION
**Pull request by rajansandeep**
Steps to follow after a CoreDNS release to update the image/manifests for CoreDNS in Kubernetes.

### CODE REVIEW

**Comment by rajansandeep** [comment]
@chrisohaver @fturib @miekg

**Comment by fturib** [comment]
Great check list !

File: kubernetes/README.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      Clearly title the issue that this is a test PR, not intended for merge. *Do not merge*.
+    * If the Kubernetes pre-submit e2e tests pass, then close the above kubernetes/kubernetes PR, and merge the coredns/deployment PR.
+
+2. Open an issue in the [Kubernetes main repository](https://github.com/kubernetes/kubernetes) to get the CoreDNS image pushed to the gcr.io repository. 
+Example issue opened for previous releases: https://github.com/kubernetes/kubernetes/issues/69983
+
+3. Once the gcr.io image has been pushed, open a PR in the [Kubernetes main repository](https://github.com/kubernetes/kubernetes) to update the CoreDNS image for both [kube-up](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/dns/coredns) and [kubeadm](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/constants/constants.go#L309) install tools.  
+    * Ensure that all the pre-submit e2e tests are passing.
+    * If there are any manifest updates from the main [CoreDNS deployment](https://github.com/coredns/deployment/blob/master/kubernetes/coredns.yaml.sed), include them in the PR.
```
Comment (chrisohaver) [inline kubernetes/README.md]: This step seems out of order with the step above it.

File: kubernetes/README.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+
+1. Update the [CoreDNS deployment](https://github.com/coredns/deployment/blob/master/kubernetes/coredns.yaml.sed) for the latest release version and ensure all the CI tests pass. But do not merge yet.
+    * Open a PR in the [Kubernetes main repository](https://github.com/kubernetes/kubernetes) that tests the new image (in Kubernetes e2e) before it is pushed to gcr.io. 
+      Clearly title the issue that this is a test PR, not intended for merge. *Do not merge*.
+    * If the Kubernetes pre-submit e2e tests pass, then close the above kubernetes/kubernetes PR, and merge the coredns/deployment PR.
+
+2. Open an issue in the [Kubernetes main repository](https://github.com/kubernetes/kubernetes) to get the CoreDNS image pushed to the gcr.io repository. 
+Example issue opened for previous releases: https://github.com/kubernetes/kubernetes/issues/69983
+
```
Comment (fturib) [inline kubernetes/README.md:13]: I guess we should say here that we need to assign to a Google person that has permission to push to gcr.io. By default, it is Tim Hockin, SIG-Network.

File: kubernetes/README.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
+# Kubernetes CoreDNS Update Release Process
+
+This document helps define the steps to be followed after a CoreDNS release to update the image/manifests for CoreDNS in the Kubernetes (and related) projects.
+After a [CoreDNS release](https://github.com/coredns/release) has been completed by one of the maintainers:
+
+
+1. Update the [CoreDNS deployment](https://github.com/coredns/deployment/blob/master/kubernetes/coredns.yaml.sed) for the latest release version and ensure all the CI tests pass. But do not merge yet.
+    * Open a PR in the [Kubernetes main repository](https://github.com/kubernetes/kubernetes) that tests the new image (in Kubernetes e2e) before it is pushed to gcr.io.
```
Comment (chrisohaver) [inline kubernetes/README.md]: I think this should be a separate step (not a sub bullet under 1).

File: kubernetes/README.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
+# Kubernetes CoreDNS Update Release Process
+
+This document helps define the steps to be followed after a CoreDNS release to update the image/manifests for CoreDNS in the Kubernetes (and related) projects.
+After a [CoreDNS release](https://github.com/coredns/release) has been completed by one of the maintainers:
+
+
+1. Update the [CoreDNS deployment](https://github.com/coredns/deployment/blob/master/kubernetes/coredns.yaml.sed) for the latest release version and ensure all the CI tests pass. But do not merge yet.
+    * Open a PR in the [Kubernetes main repository](https://github.com/kubernetes/kubernetes) that tests the new image (in Kubernetes e2e) before it is pushed to gcr.io. 
+      Clearly title the issue that this is a test PR, not intended for merge. *Do not merge*.
+    * If the Kubernetes pre-submit e2e tests pass, then close the above kubernetes/kubernetes PR, and merge the coredns/deployment PR.
```
Comment (chrisohaver) [inline kubernetes/README.md]: I think this should be a separate step (not a sub bullet under 1).

File: kubernetes/README.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+2. Open an issue in the [Kubernetes main repository](https://github.com/kubernetes/kubernetes) to get the CoreDNS image pushed to the gcr.io repository. 
+Example issue opened for previous releases: https://github.com/kubernetes/kubernetes/issues/69983
+
+3. Once the gcr.io image has been pushed, open a PR in the [Kubernetes main repository](https://github.com/kubernetes/kubernetes) to update the CoreDNS image for both [kube-up](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/dns/coredns) and [kubeadm](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/constants/constants.go#L309) install tools.  
+    * Ensure that all the pre-submit e2e tests are passing.
+    * If there are any manifest updates from the main [CoreDNS deployment](https://github.com/coredns/deployment/blob/master/kubernetes/coredns.yaml.sed), include them in the PR. 
+    * If the tests are passing, get the PR merged.
```
Comment (fturib) [inline kubernetes/README.md:25]: get the PR merged by requiring needed reviews and approvals.

File: kubernetes/README.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
+# Kubernetes CoreDNS Update Release Process
+
+This document helps define the steps to be followed after a CoreDNS release to update the image/manifests for CoreDNS in the Kubernetes (and related) projects.
+After a [CoreDNS release](https://github.com/coredns/release) has been completed by one of the maintainers:
+
+
+1. Update the [CoreDNS deployment](https://github.com/coredns/deployment/blob/master/kubernetes/coredns.yaml.sed) for the latest release version and ensure all the CI tests pass. But do not merge yet.
+    * Open a PR in the [Kubernetes main repository](https://github.com/kubernetes/kubernetes) that tests the new image (in Kubernetes e2e) before it is pushed to gcr.io. 
+      Clearly title the issue that this is a test PR, not intended for merge. *Do not merge*.
```
Comment (chrisohaver) [inline kubernetes/README.md]: ... and do a `/hold` so the PR does not merge by mistake.

File: kubernetes/README.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+- [ ] kubernetes/kubernetes passes e2e dns pre-submit conformance tests
+- [ ] Close the temporary kubernetes/kubernetes PR
+- [ ] Merge the coredns/deployment PR
+- [ ] Open a kubernetes/kubernetes issue to push CoreDNS image to k8s.gcr.io
+- [ ] Open a kubernetes/kubernetes PR to update manifests and the (k8s.gcr.io/coredns:version) image
+- [ ] kubernetes/kubernetes update PR merged
+- [ ] Verify that gce-scale-performance test passes (runs once daily)
+- [ ] Open PRs for minikube & kops
+
+
```
Comment (fturib) [inline kubernetes/README.md:48]: Great idea this checklist.

**Comment by miekg** [comment]
Can we automate it?

**Review by fturib — COMMENTED** [review]
I propose some tuning ...

File: kubernetes/README.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+2. Open an issue in the [Kubernetes main repository](https://github.com/kubernetes/kubernetes) to get the CoreDNS image pushed to the gcr.io repository. 
+Example issue opened for previous releases: https://github.com/kubernetes/kubernetes/issues/69983
+
+3. Once the gcr.io image has been pushed, open a PR in the [Kubernetes main repository](https://github.com/kubernetes/kubernetes) to update the CoreDNS image for both [kube-up](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/dns/coredns) and [kubeadm](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/constants/constants.go#L309) install tools.  
+    * Ensure that all the pre-submit e2e tests are passing.
+    * If there are any manifest updates from the main [CoreDNS deployment](https://github.com/coredns/deployment/blob/master/kubernetes/coredns.yaml.sed), include them in the PR. 
+    * If the tests are passing, get the PR merged.
+
+
+4. Validate the [5k node test](https://k8s-testgrid.appspot.com/sig-scalability-gce#gce-scale-performance) to see if the Load and Density scale tests are green.
```
Comment (chrisohaver) [inline kubernetes/README.md]: Note here that this test will not run until the PR is merged. And that it only runs once per day.   i.e. Need to be sure that the actually included merged PR.

File: kubernetes/README.md (hunk 1 above)
Comment (chrisohaver) [inline kubernetes/README.md]: > If there are any manifest updates from the main [CoreDNS deployment]

Isn't this, by definition, always the case?  Do we need to say this?

File: kubernetes/README.md (hunk 7 above)
Comment (rajansandeep) [reply kubernetes/README.md:48]: Thank you! That was @chrisohaver's suggestion.

File: kubernetes/README.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+5. Once the gcr.io image has been pushed, open a PR in the [Kubernetes main repository](https://github.com/kubernetes/kubernetes) to update the CoreDNS image 
+for both [kube-up](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/dns/coredns) and [kubeadm](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/constants/constants.go#L309) install tools.
+    * If there are any manifest updates apart from the version update from the main [CoreDNS deployment](https://github.com/coredns/deployment/blob/master/kubernetes/coredns.yaml.sed), include them in the PR. 
+    * Ensure that all the pre-submit e2e tests are passing. 
+    * If the tests are passing, get the PR merged.
+
+
+6. Validate the [5k node test](https://k8s-testgrid.appspot.com/sig-scalability-gce#gce-scale-performance) to see if the Load and Density scale tests are green. 
+    * The 5k node test tests only runs once per day. Hence ensure that tests run include the merged PR.
```
Comment (fturib) [inline kubernetes/README.md]: 5k nodes scale test ... (I guess)

**Comment by rajansandeep** [comment]
>Can we automate it?

@miekg I don't think it's possible as most steps require PRs/issues to be manually opened and updating the manifest in different places.

File: kubernetes/README.md (hunk 9 above)
Comment (fturib) [reply kubernetes/README.md]: "Hence ensure that tests run include the merged PR" - needs a rephrasing ? or maybe there is a another word for "run" here ?

**Review by fturib — COMMENTED** [review]
just few other very detailed points ...

**Review by fturib — APPROVED** [review]

**Comment by fturib** [comment]
@miekg : I guess you are the one that can merge here.

**Comment by miekg** [comment]
@fturib that would be odd, is this falls under the coredns org

**Pull request merged**